apollo-ios-codegen/Sources/GraphQLCompiler/JavaScript/src/utilities/typePolicyResolver.ts

```diff
@@ -0,0 +1,140 @@
+import {
+  DirectiveNode,
+  GraphQLCompositeType,
+  GraphQLError,
+  GraphQLInterfaceType,
+  GraphQLObjectType,
+  GraphQLSchema,
+  isInterfaceType,
+  isNonNullType,
+  isScalarType,
+  isUnionType,
+  Kind,
+} from "graphql";
+import { directive_typePolicy } from "./apolloCodegenSchemaExtension";
+
+const directiveName = directive_typePolicy.name.value;
+
+type TypePolicyDirectiveResult = {
+  directive: DirectiveNode;
+  source: GraphQLObjectType | GraphQLInterfaceType;
+};
+
+export function typePolicyDirectiveFor(
+  type: GraphQLCompositeType
+): TypePolicyDirectiveResult | undefined {
+  if (isUnionType(type)) {
+    return undefined;
+  }
+
+  for (const extension of type.extensionASTNodes ?? []) {
+    const directive = extension.directives?.find(
+      (d) => d.name.value === directiveName
+    );
+    if (directive) {
+      return { directive, source: type };
+    }
+  }
+
+  for (const directive of type.astNode?.directives ?? []) {
+    if (directive.name.value === directiveName) {
+      return { directive, source: type };
+    }
+  }
+
+  if ("getInterfaces" in type) {
+    for (const interfaceType of type.getInterfaces()) {
+      const result = typePolicyDirectiveFor(interfaceType);
+      if (result) {
+        return result;
+      }
+    }
+  }
+
+  return undefined;
+}
+
+function validateKeyFields(
+  result: TypePolicyDirectiveResult,
+  keyFields: string[]
+) {
+  const { directive, source: type } = result;
+
+  if (isUnionType(type)) {
+    return;
+  }
+
+  const label = isInterfaceType(type) ? "interface" : "object";
+
+  var allFields = type.getFields();
+  for (const keyField of keyFields) {
+    if (!keyField) {
+      throw new GraphQLError(
+        `Key fields must be a space-separated list of identifiers.`,
+        { nodes: directive }
+      );
+    }
+
+    const actualField = allFields[keyField];
+    if (!actualField) {
+      throw new GraphQLError(
+        `Key field "${keyField}" does not exist on ${label} "${type.name}".`,
+        { nodes: type.astNode ? [type.astNode, directive] : directive }
+      );
+    }
+
+    const fieldType = isNonNullType(actualField.type)
```

Comment:
@martinbonnin Do `@typePolicy` directives in Apollo Kotlin allow you to use a nullable field as a `keyField`? Just want to make sure we have the same behavior for these.

Reply:
Currently yes. Although that's a good point and we might want to restrict key fields to non-nullable in the future.

Reply:
Non-nullable has my preference as well.

The latest stable versions don't allow re-using queries with a `@typePolicy` between Apollo Kotlin and Apollo iOS anyway, so any changes in this PR won't break current usage. I could change it to require non-nullable fields here, and Apollo Kotlin could raise a warning for nullable fields.

Reply:
I've added a requirement for all key fields to be non-nullable.